near/receiver/src/governance.rs

```diff
@@ -0,0 +1,385 @@
+        data.write_u32::<BigEndian>(magic).unwrap();
+        data.push(GOVERNANCE_MODULE);
+        data.push(self.id() as u8);
+        data.extend_from_slice(&0u16.to_le_bytes());
+
+        match self {
+            Self::ContractUpgrade(hash) => {
+                data.extend_from_slice(hash);
+            }
+
+            Self::SetDataSources(sources) => {
+                data.push(sources.len() as u8);
+                for source in sources {
+                    data.extend_from_slice(&source.emitter);
+                    data.extend_from_slice(&source.pyth_emitter_chain.0.to_le_bytes());
+                }
+            }
+
+            Self::SetGovernanceSource(source) => {
+                data.extend_from_slice(&source.emitter);
+                data.extend_from_slice(&source.pyth_emitter_chain.0.to_le_bytes());
+            }
+
+            Self::SetStalePriceThreshold(stale_price_threshold) => {
+                data.extend_from_slice(&stale_price_threshold.to_le_bytes());
+            }
+
+            Self::SetUpdateFee(update_fee) => {
+                data.extend_from_slice(&update_fee.to_le_bytes());
+            }
+        }
+
+        data
+    }
+}
+
+#[near_bindgen]
+impl Pyth {
+    /// Instruction for processing Governance VAA's relayed via Wormhole.
+    ///
+    /// Note that VAA verification requires calling Wormhole so processing of the VAA itself is
+    /// done in a callback handler, see `process_vaa_callback`.
+    #[payable]
+    #[handle_result]
+    pub fn execute_governance_instruction(&mut self, vaa: String) -> Result<(), Error> {
+        // Verify the VAA is coming from a trusted source chain before attempting to verify VAA
+        // signatures. Avoids a cross-contract call early.
+        {
+            let vaa = hex::decode(&vaa).map_err(|_| Error::InvalidHex)?;
+            let vaa = serde_wormhole::from_slice_with_payload::<wormhole::Vaa<()>>(&vaa);
+            let vaa = vaa.map_err(|_| Error::InvalidVaa)?;
+            let (vaa, _rest) = vaa;
+
+            // Convert to local VAA type to catch APi changes.
+            let vaa = Vaa::from(vaa);
+
+            // Prevent VAA re-execution.
+            if self.executed_gov_sequences.contains(&vaa.sequence) {
+                return Err(Error::VaaVerificationFailed);
+            }
+
+            // Confirm the VAA is coming from a trusted source chain.
+            if self.gov_source
+                != (Source {
+                    emitter:            vaa.emitter_address,
+                    pyth_emitter_chain: vaa.emitter_chain,
+                })
+            {
+                return Err(Error::UnknownSource);
+            }
+
+            // Insert before calling Wormhole to prevent re-execution. If we wait until after the
+            // Wormhole call we could end up with multiple VAA's with the same sequence being
+            // executed in parallel.
+            self.executed_gov_sequences.insert(&vaa.sequence);
```

Comment:
I think with the current implementation someone can just deny our executed governance sequences by calling this with fake vaa. Can you explain why it can cause re-execution if we have it after wormhole call? this contract won't be executed parallel, right?

Reply:
Also it's good to use a number instead of set as we do in other chains.. because we do not want out of order vaa execution.